sql/catalog/tabledesc: TestRemoveDefaultExprFromComputedColumn failed [timed out while starting tenant]

sql/catalog/tabledesc.TestRemoveDefaultExprFromComputedColumn [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/12603206?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/12603206?buildTab=artifacts#/) on master @ [d3f1f25c822bdb282d250834cae1f7b3475142d7](https://github.com/cockroachdb/cockroach/commits/d3f1f25c822bdb282d250834cae1f7b3475142d7):

Fatal error:

```
panic: test timed out after 57s
running tests:
	TestRemoveDefaultExprFromComputedColumn (6s)

```
Stack: 

```
goroutine 30130 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2259 +0x320
created by time.goFunc
	GOROOT/src/time/sleep.go:176 +0x38
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestRemoveDefaultExprFromComputedColumn
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/7b4578e3c0124849ee4088d817d512f1/logTestRemoveDefaultExprFromComputedColumn3327043814
    test_log_scope.go:81: use -show-logs to present logs inline
    test_server_shim.go:159: automatically injected a shared process virtual cluster under test; see comment at top of test_server_shim.go for details.
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRemoveDefaultExprFromComputedColumn.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-33394

--- a/pkg/base/config.go
+++ b/pkg/base/config.go
@@ -885,8 +885,23 @@ func TempStorageConfigFromEnv(
 	maxSizeBytes int64,
 ) TempStorageConfig {
 	inMem := parentDir == "" && useStore.InMemory
+	return newTempStorageConfig(ctx, st, inMem, useStore, maxSizeBytes)
+}
+
+// InheritTempStorageConfig creates a new TempStorageConfig using the
+// configuration of the given TempStorageConfig. It assumes the given
+// TempStorageConfig has been fully initialized.
+func InheritTempStorageConfig(
+	ctx context.Context, st *cluster.Settings, parentConfig TempStorageConfig,
+) TempStorageConfig {
+	return newTempStorageConfig(ctx, st, parentConfig.InMemory, parentConfig.Spec, parentConfig.Mon.Limit())
+}
+
+func newTempStorageConfig(
+	ctx context.Context, st *cluster.Settings, inMemory bool, useStore StoreSpec, maxSizeBytes int64,
+) TempStorageConfig {
 	var monitorName redact.RedactableString
-	if inMem {
+	if inMemory {
 		monitorName = "in-mem temp storage"
 	} else {
 		monitorName = "temp disk storage"
@@ -902,7 +917,7 @@ func TempStorageConfigFromEnv(
 	)
 	monitor.Start(ctx, nil /* pool */, mon.NewStandaloneBudget(maxSizeBytes))
 	return TempStorageConfig{
-		InMemory: inMem,
+		InMemory: inMemory,
 		Mon:      monitor,
 		Spec:     useStore,
 		Settings: st,

--- a/pkg/base/test_server_args.go
+++ b/pkg/base/test_server_args.go
@@ -526,6 +526,7 @@ func DefaultTestTempStorageConfigWithSize(
 	return TempStorageConfig{
 		InMemory: true,
 		Mon:      monitor,
+		Spec:     DefaultTestStoreSpec,
 		Settings: st,
 	}
 }
@@ -548,6 +549,7 @@ func InheritTestTempStorageConfig(
 	return TempStorageConfig{
 		InMemory: parentConfig.InMemory,
 		Path:     parentConfig.Path,
+		Spec:     parentConfig.Spec,
 		Mon:      monitor,
 		Settings: st,
 	}

--- a/pkg/base/test_server_args.go
+++ b/pkg/base/test_server_args.go
@@ -531,30 +531,6 @@ func DefaultTestTempStorageConfigWithSize(
 	}
 }
 
-// InheritTestTempStorageConfig is the associated temp storage for
-// a secondary tenant.
-func InheritTestTempStorageConfig(
-	st *cluster.Settings, parentConfig TempStorageConfig,
-) TempStorageConfig {
-	monitor := mon.NewMonitor(
-		"in-mem temp storage",
-		mon.DiskResource,
-		nil,                                    /* curCount */
-		nil,                                    /* maxHist */
-		1024*1024,                              /* increment */
-		DefaultInMemTempStorageMaxSizeBytes/10, /* noteworthy */
-		st,
-	)
-	monitor.Start(context.Background(), parentConfig.Mon, nil /* reserved */)
-	return TempStorageConfig{
-		InMemory: parentConfig.InMemory,
-		Path:     parentConfig.Path,
-		Spec:     parentConfig.Spec,
-		Mon:      monitor,
-		Settings: st,
-	}
-}
-
 // TestSharedProcessTenantArgs are the arguments to
 // TestServer.StartSharedProcessTenant.
 type TestSharedProcessTenantArgs struct {

--- a/pkg/server/server_controller_new_server.go
+++ b/pkg/server/server_controller_new_server.go
@@ -321,15 +321,18 @@ func makeSharedProcessTenantServerConfig(
 	useStore := kvServerCfg.SQLConfig.TempStorageConfig.Spec
 	tempStorageCfg := base.TempStorageConfigFromEnv(
 		ctx, st, useStore, "" /* parentDir */, kvServerCfg.SQLConfig.TempStorageConfig.Mon.Limit())
-	// TODO(knz): Make tempDir configurable.
-	tempDir := useStore.Path
-	if tempStorageCfg.Path, err = fs.CreateTempDir(tempDir, TempDirPrefix, stopper); err != nil {
-		return BaseConfig{}, SQLConfig{}, errors.Wrap(err, "could not create temporary directory for temp storage")
-	}
-	if useStore.Path != "" {
-		recordPath := filepath.Join(useStore.Path, TempDirsRecordFilename)
-		if err := fs.RecordTempDir(recordPath, tempStorageCfg.Path); err != nil {
-			return BaseConfig{}, SQLConfig{}, errors.Wrap(err, "could not record temp dir")
+
+	if !tempStorageCfg.InMemory {
+		// TODO(knz): Make tempDir configurable.
+		tempDir := useStore.Path
+		if tempStorageCfg.Path, err = fs.CreateTempDir(tempDir, TempDirPrefix, stopper); err != nil {
+			return BaseConfig{}, SQLConfig{}, errors.Wrap(err, "could not create temporary directory for temp storage")
+		}
+		if useStore.Path != "" {
+			recordPath := filepath.Join(useStore.Path, TempDirsRecordFilename)
+			if err := fs.RecordTempDir(recordPath, tempStorageCfg.Path); err != nil {
+				return BaseConfig{}, SQLConfig{}, errors.Wrap(err, "could not record temp dir")
+			}
 		}
 	}
 

--- a/pkg/server/server_controller_new_server.go
+++ b/pkg/server/server_controller_new_server.go
@@ -318,11 +318,9 @@ func makeSharedProcessTenantServerConfig(
 		baseCfg.InflightTraceDirName = traceDir
 	}
 
-	useStore := kvServerCfg.SQLConfig.TempStorageConfig.Spec
-	tempStorageCfg := base.TempStorageConfigFromEnv(
-		ctx, st, useStore, "" /* parentDir */, kvServerCfg.SQLConfig.TempStorageConfig.Mon.Limit())
-
+	tempStorageCfg := base.InheritTempStorageConfig(ctx, st, kvServerCfg.SQLConfig.TempStorageConfig)
 	if !tempStorageCfg.InMemory {
+		useStore := tempStorageCfg.Spec
 		// TODO(knz): Make tempDir configurable.
 		tempDir := useStore.Path
 		if tempStorageCfg.Path, err = fs.CreateTempDir(tempDir, TempDirPrefix, stopper); err != nil {

--- a/pkg/server/testserver.go
+++ b/pkg/server/testserver.go
@@ -611,7 +611,7 @@ func (ts *testServer) startDefaultTestTenant(
 
 	var tempStorageConfig base.TempStorageConfig
 	if tsc := ts.params.TempStorageConfig; tsc.Settings != nil {
-		tempStorageConfig = base.InheritTestTempStorageConfig(tenantSettings, tsc)
+		tempStorageConfig = base.InheritTempStorageConfig(ctx, tenantSettings, tsc)
 	} else {
 		tempStorageConfig = base.DefaultTestTempStorageConfig(tenantSettings)
 	}

--- a/pkg/sql/colflow/draining_test.go
+++ b/pkg/sql/colflow/draining_test.go
@@ -53,7 +53,7 @@ func TestDrainingAfterRemoteError(t *testing.T) {
 	diskMonitor.Start(ctx, nil /* pool */, mon.NewStandaloneBudget(1))
 
 	// Set up a two node cluster.
-	tempStorageConfig := base.TempStorageConfig{InMemory: true, Mon: diskMonitor, Settings: st}
+	tempStorageConfig := base.TempStorageConfig{InMemory: true, Mon: diskMonitor, Settings: st, Spec: base.DefaultTestStoreSpec}
 	args := base.TestClusterArgs{
 		ServerArgs: base.TestServerArgs{
 			Settings:          st,

